File: task2/.js
function normalizeEmailJs(email) {
    const parts = email.split('@'); // Разбиваем email на имя и домен
    if (parts.length !== 2) { // Проверяем валидность (есть @?)
        return ""; // Некорректный формат
    }
    let username = parts[0]; // Имя пользователя
    const domain = parts[1]; // Домен

    let normalizedUsername = ""; // Для нормализованного имени
    for (let char of username) { // Перебираем символы имени
        if (char === '.') { // Пропускаем точки
            continue;
        } else if (char === '+' || char === '*') { // Прерываем после + или *
            break;
        } else {
            normalizedUsername += char; // Добавляем символы
        }
    }
    return normalizedUsername + "@" + domain; // Возвращаем нормализованный email
}

function countUniqueEmailsJs(emails) {
    const uniqueEmails = new Set(); // Set для уникальных email
    for (const email of emails) { // Перебираем все email
        uniqueEmails.add(normalizeEmailJs(email)); // Нормализуем и добавляем
    }
    return uniqueEmails.size; // Возвращаем кол-во уникальных
}

const emails1 = ["mar.pha+science@corp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"];
console.log("Example 1 Output:", countUniqueEmailsJs(emails1));
const emails2 = ["mar.pha+science@co.rp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"];
console.log("Example 2 Output:", countUniqueEmailsJs(emails2));

File: task2/.go
package main
import (
	"fmt"       
	"strings" 
)

func normalizeEmailGo(email string) string {
	parts := strings.Split(email, "@") 
	if len(parts) != 2 {              // Проверяем, что email имеет @
		return "" 
	}
	username := parts[0]   
	domain := parts[1]     

	normalizedUsername := "" 
	for _, char := range username {  // Перебираем символы имени пользователя
		if char == '.' {        
			continue
		} else if char == '+' || char == '*' { 
			break
		} else {
			normalizedUsername += string(char) // Добавляем символ к нормализованному имени
		}
	}
	return normalizedUsername + "@" + domain // Возвращаем нормализованный email
}

func countUniqueEmailsGo(emails []string) int {
	uniqueEmails := make(map[string]bool) 
	for _, email := range emails {         // Перебираем все email
		uniqueEmails[normalizeEmailGo(email)] = true // Нормализуем и добавляем в map
	}
	return len(uniqueEmails) 
}

func main() {
	emails1 := []string{"mar.pha+science@corp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"} 
	fmt.Println("Example 1 Output:", countUniqueEmailsGo(emails1))                   

	emails2 := []string{"mar.pha+science@co.rp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"} 
	fmt.Println("Example 2 Output:", countUniqueEmailsGo(emails2))                   
}
